pywr-schema/tests/test_models.rs

```diff
@@ -0,0 +1,143 @@
+#[cfg(feature = "core")]
+use pywr_core::test_utils::{run_all_solvers, ExpectedOutputs};
+use pywr_schema::PywrModel;
+use std::fs;
+use std::path::Path;
+#[cfg(feature = "core")]
+use std::path::PathBuf;
+#[cfg(feature = "core")]
+use tempfile::TempDir;
+
+macro_rules! model_tests {
+    ($($test_func:ident: $value:expr,)*) => {
+    $(
+        #[test]
+        fn $test_func() {
+
+            // Deserialise the schema and run it
+            #[cfg(feature = "core")]
+            {
+                let (input, expected, solvers_without_features): (&str, Vec<&str>, Vec<&str>) = $value;
+                let input_pth = Path::new(env!("CARGO_MANIFEST_DIR")).join("tests").join(input);
+                let expected_paths = expected.iter().map(|p| Path::new(env!("CARGO_MANIFEST_DIR")).join("tests").join(p)).collect::<Vec<_>>();
+                let schema = deserialise_test_model(&input_pth);
+                run_test_model(&schema, &expected_paths, &solvers_without_features);
+            }
+
+            // Just deserialise the schema
+            #[cfg(not(feature = "core"))]
+            {
+                let (input, _expected, _solvers_without_features): (&str, Vec<&str>, Vec<&str>) = $value;
+                let input_pth = Path::new(env!("CARGO_MANIFEST_DIR")).join("tests").join(input);
+                let _schema = deserialise_test_model(&input_pth);
+            }
+        }
+    )*
+    }
+}
+
+model_tests! {
+    test_simple1: ("simple1.json", vec![], vec![]),
+    test_csv1: ("csv1.json", vec!["csv1-outputs-long.csv", "csv1-outputs-wide.csv"], vec![]),
+    test_csv2: ("csv2.json", vec!["csv2-outputs-long.csv", "csv2-outputs-wide.csv"], vec![]),
+    test_csv3: ("csv3.json", vec!["csv3-outputs-long.csv"], vec![]),
+    test_hdf1: ("hdf1.json", vec![], vec![]), // TODO asserting h5 results not possible with this framework
+    test_memory1: ("memory1.json", vec![], vec![]),  // TODO asserting memory results not possible with this framework
```

Comment:
I assume these will need custom testing functions in this script and this will be the case for any tests that cannot be done by checking csv outputs?

Reply:
In this case I left the unit tests in the modules, but included them here to test deserialisation.